Refactor synchronicity matrix new

colombe-tolokin opened the pull request:
- [x] fix #11 
- [x] description of feature/fix

# Refactored Complete Synchronicity Matrix System - Completed

### **Matrix Creation** (`R/synch_matrix_creation.R`)
- `create_time_sequence` - Creates time sequences from start to end by seconds
- `prepare_time_animal_matrix` - Creates empty time-animal matrices for synchronicity analysis  
- `prepare_time_bin_matrix` - Creates empty time-bin matrices for tracking bin occupancy
- `prepare_time_feed_matrix` - Creates empty time-feed matrices for feed bin tracking

### **Matrix Processing** (`R/synch_matrix_processing.R`)
- `empty_synch_matrix` - Generates empty synchronization matrices for feed/water data
- `matrix_initialize` - Initializes and processes synchronization matrices
- `process_cur_synch` - Processes synchronization data to replace NA values and compute totals
- `matrix_process` - Processes matrices and adds derived columns

### **Feed/Drink Combined Analysis** (`R/synch_feed_drink_combined.R`)
- `total_animals_present` - Calculates total animals present at each time point
- `delete_inactive_time` - Removes time periods when no animals are active
- `add_date` - Adds date columns to synchronicity matrices
- `bin_update` - Updates bin numbers for feed/water mapping
- `feed_drink_matrix_process` - Main workflow for combined feed/drink matrix processing

### **Synchronicity Analysis** (`R/synch_synchronicity_analysis.R`)
- `empty_animal_matrix` - Creates empty Animal × Animal matrices for analysis
- `calculate_bout_duration` - Calculates bout numbers and durations for time series
- `paired_synchronicity_analysis` - Analyzes paired animal synchronicity (simultaneous activity)
- `neighbor_synchronicity_analysis` - Analyzes neighboring animal synchronicity (adjacent bins)
- `synchronicity_matrix_process` - Main orchestrator for complete synchronicity analysis

**Total: 18 functions refactored across 4 files**

- [x] tests added/passed
- [ ] add an entry to the [changelog](../CHANGELOG.md)

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduce a full synchronicity analysis workflow (matrix creation/processing, pair and neighbor analyses) with a new configurable physical bin layout, plus docs/tests and pkg site updates.
> 
> - **Synchronicity Analysis (new)**:
>   - Add time-sequence/matrix builders in `R/synch_matrix_creation.R` (`create_time_sequence`, `prepare_time_animal_matrix`, `prepare_time_feed_matrix`, `prepare_time_bin_matrix`).
>   - Add processing pipeline in `R/synch_matrix_processing.R` (`matrix_process` + internal helpers) to populate/filter matrices and compute derived columns.
>   - Add analyses:
>     - Pair co-occurrence: `synch_pair_analysis` (`R/synch_pair_analysis.R`).
>     - Spatial neighbor co-occurrence using physical layout: `synch_neighbor_analysis` (`R/synch_neighbor_analysis.R`).
>   - Utilities in `R/synch_utils.R` (pair matrices, bout stats, bin-layout parsing/neighbors).
> - **Configuration (global vars)**:
>   - New physical bin layout global (`bin_layout2`, `set_bin_layout2`), integrated into `set_global_cols()`; default layout added in `R/global_var.R`.
> - **Exports/Dependencies**:
>   - Export new APIs: `matrix_process`, `synch_pair_analysis`, `synch_neighbor_analysis`, `bin_layout2`, `set_bin_layout2`.
>   - Add `zoo` to `Imports`.
> - **Docs/Website**:
>   - New man pages for added functions; `_pkgdown.yml` adds “Synchronicity Analysis” section and references new APIs.
>   - Vignette `vignettes/data_cleaning.Rmd` now documents `bin_layout` usage.
>   - README: add contributor ORCID.
> - **Tests**:
>   - Extensive `tests/testthat/` covering matrix creation/processing, pair/neighbor analysis, utils, and new global config.
> - **Repo/Meta**:
>   - Update `.gitignore`, `.Rbuildignore` entries (e.g., `pkgdown/`, `Icon`, `doc`, `Meta`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f0155982d94ce4ae07e13c08b6836d9103fbdd4d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

colombe-tolokin committed:
refactor: add matrix creation logic and tests (create_time_sequence, prepare_time_cow_matrix, prepare_time_bin_matrix, prepare_time_feed_matrix)

colombe-tolokin committed:
refactor: add matrix processing logic and tests (empty_synch_matrix, matrix_initialize, process_cur_synch, matrix_process)

colombe-tolokin committed:
fix: remove problematic @docType package from matrix creation documentation

colombe-tolokin committed:
fix: add zoo dependency to DESCRIPTION to resolve CI warnings

colombe-tolokin committed:
fix: mark internal helper functions as non-exported to resolve pkgdown build issues

colombe-tolokin committed:
fix: configure pkgdown to include synchronicity matrix functions and mark documentation topics as internal

colombe-tolokin committed:
fix: robust time matching, type-conditional logic, and NA fill for matrix processing

skysheng7 reviewed (changes requested):
Thank you @colombe-tolokin for the excellent work on refactoring the synchronicity matrix code! I really appreciate your fast response and working so hard on this function. I know it's a very long and complicated script to refactor, so thank you for taing on the hard challenge.

Below are some of my feedback:

## 🚫 **Critical Issues to Address**

Here are some important issues that need to be fixed before merging. The main problems is about not using the global variables and inconsistency in code style.

### 1. **Hard-coded Column Names**
The code uses hard-coded column names instead of our global variable system:

**Current (❌):**
```r
if (!all(c("Start", "End") %in% names(cur_data))) stop("cur_data must have 'Start' and 'End' columns")
if (!("Cow" %in% names(cur_data))) stop("cur_data must have a 'Cow' column")
cur_data <- cur_data[order(cur_data$Start, cur_data$End), ]
```

**Should be (✅):**
```r
required_cols <- c(start_col, end_col, id_col, bin_col, start_weight_col, end_weight_col)
missing_cols <- setdiff(required_cols, names(cur_data))
if (length(missing_cols) > 0) {
  stop("Missing required columns: ", paste(missing_cols, collapse = ", "))
}
cur_data <- cur_data[order(cur_data[[start_col]], cur_data[[end_col]]), ]
```

### 2. **Hard-coded Bin Parameters**
Functions require manual specification of bin ranges instead of using our global variables:

**Current (❌):**
```r
matrix_process(data_list, min_feed_bin = 1, max_feed_bin = 30, total_feed_bin = 30)
```

**Should be (✅):**
```r
matrix_process(data_list, bins_feed = bins_feed2(), bins_wat = bins_wat2())
```

Then you can extract the min_feed_bin, max_feed_bin, total_feed_bin using bins_feed. This is to make sure we always use global variables.

### 3. **Test Files: Incorrect Package Loading**
Your test files use `library()` calls which violate R package development standards:

**Current (❌):**
```r
library(testthat)
library(lubridate)
library(zoo)
```

- We will never use `library()` to load packages during R package development, all packages we use should be listed in DESCRIPTION, and will be auto-loaded.

**Should be (✅):**
```r
# Remove all library() calls
# Use package::function() notation instead:
test_that("create_time_sequence works for normal input", {
  df <- data.frame(Start = lubridate::ymd_hms("2023-01-01 00:00:00"), 
                   End = lubridate::ymd_hms("2023-01-01 00:00:05"))
  # ... rest of test
})
```

### 4. **Function Signatures Need Global Variable Integration**
All functions should follow the pattern established in `R/cluster_meals.R`:

- For example:

```r
function_name <- function(data, 
                         id_col = id_col2(),
                         start_col = start_col2(),
                         end_col = end_col2(),
                         bin_col = bin_col2(),
                         start_weight_col = start_weight_col2(),
                         end_weight_col = end_weight_col2(),
                         bins_feed = bins_feed2(),
                         bins_wat = bins_wat2(),
                         tz = tz2()) {
  # function body
}
```

### 5. **Documentation Issues**
- Use "animal" instead of "cow" in documentation (as per our generalization rule to make it about all animals not just about cows, by default the id_col is set to "cow" to give user an example, but i want them to be able to generalize the analysis to all animals)
- Add `@inheritParams set_global_cols` to avoid parameter documentation duplication
- Add working `@examples` sections for exported functions (currently missing for some)
- Follow the documentation pattern from `R/cluster_meals.R`

### 6. **Test Coverage Enhancement**
While you have good basic tests, we need to ensure **>95% code coverage** with tests that cover:
- **Normal use cases**: ✅ You have these
- **Edge cases**: ⚠️ Need more (empty data, single row data, etc.)
- **Error handling**: ✅ You have good error tests
- **Integration tests**: ⚠️ Need more tests with toy data examples (more rows per toy dataframe)

## 📋 **Action Items**

### **Immediate Fixes Needed:**

1. **Update function signatures** to use global variable system:
   ```r
   # Add these parameters to all functions (only if you use these column names)
   id_col = id_col2(),
   start_col = start_col2(),
   end_col = end_col2(),
   bin_col = bin_col2(),
   start_weight_col = start_weight_col2(),
   end_weight_col = end_weight_col2(),
   bins_feed = bins_feed2(),
   bins_wat = bins_wat2(),
   tz = tz2()
   ```

2. **Replace all hard-coded column references** with `data[[col_name]]` pattern

3. **Fix test files** - remove all `library()` calls and use `package::function()` notation

4. **Add proper `@inheritParams` documentation** to avoid duplication

5. **Add working `@examples` sections** for all exported functions

6. **Replace "cow" with "animal"** in all documentation

7. **Add comprehensive input validation** following our standards

### **Development Workflow Suggestions: Cursor**

- **Use Cursor Rules**: Sometimes Cursor forgot it has rules, make sure you set the rule to be always run upon every request. And sometimes you may need to reference `@my-cursor-rules.mdc` in your Cursor chat when working on this.
- **Model Selection**: please use calude-4-sonnet, it is lower cost (in terms of how many requests it consumes), and high performance with high accuracy.
- **BugBot** I realized that BugBot is account dependent not repository dependent. So although I set it up using my account, it only runs when I create the PR or make a comment, you can set it up for your account following the instructions here: https://docs.cursor.com/bugbot
- **Follow Example Pattern**: Use `@cluster_meals.R` as your template for coding style, function signatures, and documentation
- **Test Coverage**: When writing tests, remind Cursor to "write tests to cover normal, edge and error handling cases, make sure >95% code coverage"


Thank you again for your hard work on this challenging refactoring!

colombe-tolokin committed:
Refactor synchronicity matrix functions - address PR feedback: Replace hard-coded column names with global variables, update function signatures, rename prepare_time_cow_matrix() to prepare_time_animal_matrix(), update cow->animal documentation, add @examples/@inheritParams, remove library() from tests, fix BugBot NA/indexing issues, update return values synch_master_cow->synch_master_animal. All 1924 tests passing, R CMD check clean.

colombe-tolokin committed:
chore: ignore pkgdown generated files - Add pkgdown/ to .gitignore to exclude generated documentation files

colombe-tolokin committed:
refactor: convert legacy synchronicity matrix file to compatibility layer - Replace hard-coded implementations with calls to new refactored functions, add deprecation warnings, maintain backward compatibility

colombe-tolokin committed:
feat: implement  Feed & Drink Combined Synchronicity Matrix Processing

Add comprehensive feed/drink combined analysis functions :

Core Functions:

- total_animals_present(): Calculate total animals and bin occupancy metrics

- delete_inactive_time(): Filter matrices to active periods only

- add_date(): Add date columns and update list names from Time column

- bin_update(): Update bin numbers for combined feed/water analysis

- feed_drink_matrix_process(): Main for complete workflow

colombe-tolokin committed:
feat: Enforce comprehensive feedback standards across synchronicity functions - Standardize 33 error messages with backtick convention across all functions - Functions 1-4 (Matrix Creation): 9 corrections, Functions 5-8 (Matrix Processing): 4 corrections - Functions 9-13 (Feed & Drink Combined): 15 corrections, Functions 14-18 (Synchronicity Analysis): 5 corrections - Align all test expectations with updated error messages - Maintain 100% Cursor rules compliance and moo4feed standards - All 370 synchronicity tests passing with 0 errors/warnings/notes

colombe-tolokin committed:
Add ORCID for Colombe Tolokin contributor

- Add ORCID 0009-0002-5355-4851 to DESCRIPTION file
- Add ORCID 0009-0002-5355-4851 to README.md Contributors section

colombe-tolokin committed:
Merge remote-tracking branch 'origin/main' into refactor-synchronicity-matrix-new

colombe-tolokin committed:
fix: PR feedback for synchronicity matrix functions

- Replace hard-coded column names with global variable system

- Remove library() calls from test files, use package::function() notation

- Change documentation examples from "cow" to "animal" for generalization

- Update examples

- NA handling and column indexing errors

- Improve test coverage for sync files

colombe-tolokin committed:
fix: update sync function examples to follow package conventions - use explicit column parameters instead of global variable setting

colombe-tolokin committed:
fix: ensure sync internal functions use dontrun pattern consistently

colombe-tolokin committed:
fix: implement feedbacks

skysheng7 committed:
remove the begining header of @synch_matrix_creation.R

skysheng7 committed:
change gitignore

skysheng7 committed:
Refactor synchronicity matrix processing

- Updated synch_feed_drink_combined.R, synch_matrix_creation.R, synch_matrix_processing.R, and synch_synchronicity_analysis.R
- Modified NAMESPACE and documentation files
- Updated test files for synchronicity matrix functions
- Removed obsolete documentation files (empty_synch_matrix.Rd, matrix_initialize.Rd)

skysheng7 committed:
checked matrix process all works

skysheng7 committed:
Merge branch 'main' into refactor-synchronicity-matrix-new

skysheng7 committed:
fix bug identified by cursor

skysheng7 committed:
do pair analysis seperately

skysheng7 committed:
fix all synch code issues

skysheng7 committed:
fix the issue of long string in md

skysheng7 committed:
tests added
